src/lsm/manifest_log.zig

```diff
@@ -954,47 +1000,237 @@ pub fn ManifestLogType(comptime Storage: type) type {
+
+        const half_bar_compact_blocks_extra = options.compact_extra_blocks;
+        assert(half_bar_compact_blocks_extra > 0);
+
+        // "C":
+        const half_bar_compact_blocks_max =
+            half_bar_append_blocks_max + half_bar_compact_blocks_extra;
+        assert(half_bar_compact_blocks_max > half_bar_append_blocks_max);
+
+        // "T":
+        const log_blocks_full_max = stdx.div_ceil(options.tables_max, block_entries_max);
+
+        // "limit of MC(c) as c approaches ∞":
+        // Working out this recurrence relation's limit with a closed-form solution is complicated.
+        // Just compute the limit iteratively instead. (1024 is an arbitrary safety counter.)
+        var log_blocks_before: u32 = 0;
+        const log_blocks_cycle_max = for (0..1024) |_| {
+            const log_blocks_after =
+                log_blocks_full_max +
+                half_bar_append_blocks_max *
+                stdx.div_ceil(log_blocks_before, half_bar_compact_blocks_max);
+
+            if (log_blocks_before == log_blocks_after) {
+                break log_blocks_after;
+            }
+            log_blocks_before = log_blocks_after;
+        } else {
+            // If the value does not converge within the given number of steps,
+            // constants.lsm_manifest_compact_blocks_extra should probably be raised.
+            @panic("ManifestLog.Pace.log_blocks_cycle_max: no convergence");
+        };
+
+        const log_blocks_burst_max = half_bar_append_blocks_max *
+            stdx.div_ceil(log_blocks_full_max + 1, half_bar_compact_blocks_max);
+
+        // "limit of MB(b) as b approaches ∞":
+        const log_blocks_max = log_blocks_cycle_max + log_blocks_burst_max;
 
-    fn blocks_count_appends(options: *const Options) usize {
-        return stdx.div_ceil(options.compaction_appends_max(), schema.ManifestNode.entry_count_max);
+        return .{
+            .half_bar_append_blocks_max = half_bar_append_blocks_max,
+            .half_bar_compact_blocks_max = half_bar_compact_blocks_max,
+            .log_blocks_full_max = log_blocks_full_max,
+            .log_blocks_max = log_blocks_max,
+            .log_blocks_cycle_max = log_blocks_cycle_max,
+        };
     }
 
-    /// The upper-bound of manifest blocks we must buffer.
-    ///
-    /// `blocks` must have sufficient capacity for:
-    /// - a manifest log compaction (+1 block in the worst case)
-    /// - a leftover open block from the previous ops (+1 block)
-    /// - table updates from a half bar of compactions
-    fn blocks_count_max(options: *const Options) usize {
-        const count = 1 + 1 + options.blocks_count_appends();
-        assert(count >= 3);
-
-        return count;
+    fn half_bar_compact_blocks(pace: Pace, options: struct {
+        /// The number of manifest blocks that *currently* exist.
+        log_blocks_count: u32,
+    }) u32 {
+        if (options.log_blocks_count < pace.log_blocks_cycle_max) {
```

Comment:
I _think_ this might want to be `pace.log_blocks_cycle_max - (half_bar_append_blocks_max)`.

If I understand the equations correctly, the invariant we prove is:

> If currently the manifest has `log_blocks_cycle_max` blocks, and you compact at max pace (pace.half_bar_compact_blocks_max), then the size of the manifest would not grow beyond `log_blocks_max`

But we might not get a chance to kick compaction into full gear exactly at `log_blocks_cycle_max` blocks! Eg, if we don't compact at all until we have `log_blocks_cycle_max`, we might be in a situation where we have exactly `log_blocks_cycle_max - 1` blocks, don't compact, and append `half_bar_append_blocks_max`. So we only start compacting from 

`log_blocks_cycle_max + half_bar_append_blocks_max` blocks, and that might hit a tiny bit higher? 


(given that the bound is conservative, it probably works out even without this adjustment, but this I think doesn't follow from our calcualtions).

Reply:
Also, at the end of the doc comment, let's formulate the postivie invariat we have:

> If anytime manifest log at least `X` long you compact at last `Y` blocks, the total size of the manifest can't be larger than `Z`

That is, on the first reading it was not obvious that what's at play here is not only the speed we compact and at the speed we append at, but also the point when we _start_ compacting.